### ISSUE TITLE
👋Welcome ms-swift v4

### ISSUE THREAD

**Issue by Jintao-Huang** [modelscope/ms-swift#7250]
ms-swift v4将会对整体框架进行重构，这会引入不少break change。以下为会导致break change的重构内容：

1. 目录结构重构与依赖关系优化：swift.llm目录将拆分为swift.template, swift.dataset, swift.model, swift.pipelines。
2. model_type与template接耦，与transformers model_type对齐，将去除qwen3_thinking, qwen3_no_thinking, qwen3_moe_thinking等额外的model_type。
3. megatron训练 training loop重写，支持RAY。使用megatron-core而丢弃使用megatron-lm依赖。
4. transformers v5.0新特性引入。

如果有更多重构/新特性支持的建议，可以在下方留言，该文档会持续更新，谢谢！

ms-swift v4 will undergo a comprehensive framework refactoring, which will introduce several breaking changes. The following are refactoring items that will cause breaking changes:

1. Directory structure refactoring and dependency optimization: The swift.llm directory will be split into swift.template, swift.dataset, swift.model, and swift.pipelines.
2. Decoupling model_type from template, aligning with transformers model_type. Additional model_types such as qwen3_thinking, qwen3_no_thinking, qwen3_moe_thinking, etc. will be removed.
3. Megatron training loop rewrite to support RAY. Will use megatron-core and drop the megatron-lm dependency.
4. Introduction of transformers v5.0 new features.

If you have more suggestions for refactoring/new feature support, please leave a comment below. This document will be continuously updated. Thank you!

**Comment by edc3000** [comment]
加入对remote的LLM-as-judge奖励的支持，不用同步等待奖励计算时间，减少GPU气泡

**Comment by chuxiliyixiaosa** [comment]
pt推理时，读取数据与模型推理是串行的，当推理图像、视频等数据时，会导致推理效率低下。

**Comment by hjh0119** [comment]
@edc3000 maybe async reward can help https://github.com/modelscope/ms-swift/pull/7252

**Comment by Jintao-Huang** [comment]
> pt推理时，读取数据与模型推理是串行的，当推理图像、视频等数据时，会导致推理效率低下。

是的 会优化一下

**Comment by aJupyter** [comment]
与其他框架RL速度的对比

**Comment by LiXinYuECNU** [comment]
使用swift推理，当某个样本数据加载报错会直接中断其他的推理

**Comment by zyjcsf** [comment]
希望可以增加对Omni模型talker部分的微调训练支持哈

**Comment by xiyu-Luoy** [comment]
hope support vllm-omni

**Comment by 651961** [comment]
希望支持一下MiMo-V2-Flash的MOPD（Multi-Teacher On-Policy Distillation），on policy distillation的方法很有潜力，希望能做的更简单易用。

**Comment by JoshonSmith** [comment]
希望可以增加对Omni模型talker部分的微调训练

**Comment by howitry** [comment]
希望可以优化流式加载数据集场景下的训练速度，目前流式加载数据集的速度不够理想：（1）不支持dataloader_num_workers>1；（2）dataloader只会在rank==0处理数据并分发到其他rank，这样训练速度会随着机器的增加而变慢，非常不利于流式加载场景下的多机多卡训练。

**Comment by hpsun1109** [comment]
希望支持一下扩散模型训练

**Comment by hpsun1109** [comment]
希望支持Nemotron-3-Nano-30B-A3B等mamba架构模型

**Comment by inzamam-iqbal** [comment]
Supporting Pluginization[](https://swift.readthedocs.io/en/latest/Customization/Pluginization.html#pluginization)  with call backs for megatron training will be very useful. Currently when megatron the training is not very customizable unless we make code changes to the ms-swift package it self. Thank you.

**Comment by Jintao-Huang** [comment]
> 希望可以优化流式加载数据集场景下的训练速度，目前流式加载数据集的速度不够理想：（1）不支持dataloader_num_workers>1；（2）dataloader只会在rank==0处理数据并分发到其他rank，这样训练速度会随着机器的增加而变慢，非常不利于流式加载场景下的多机多卡训练。

好的 这个确实很重要

**Comment by Jintao-Huang** [comment]
> 希望可以增加对Omni模型talker部分的微调训练支持哈

计划中 会在swift 4.1 支持

**Comment by huangfu170** [comment]
希望支持一个脚本进行多机训练

**Comment by Jintao-Huang** [comment]
https://github.com/modelscope/ms-swift/tree/main/examples/train/multi-node

> 希望支持一个脚本进行多机训练

**Comment by gsinthong** [comment]
Does ms-swift currently support (or plan to support) a single rollout / inference server backed by one model that is TP sharded across multiple nodes (multi-node TP inference, one logical model), specifically for [Megatron-based GRPO rollout](https://github.com/modelscope/ms-swift/blob/main/examples/megatron/grpo/dense_server.sh), similar to vLLM multi-node on Ray?

**Comment by panqingfeng** [comment]
希望能支持nvidia dgx spark机器的训练

**Comment by chizhanyuefeng** [comment]
有关强化学习方面，是否会支持verl？

**Comment by Virgil-L** [comment]
希望未来能支持量化感知训练

**Comment by Amandaynzhou** [comment]
希望优化dataloader，sft 数据量不小，多节点训练时，目前发现: 
1. streaming  效率较低
2.使用前置 mapping dataset: interleave_prob调整后会有漫长的重新mapping过程

**Comment by dongZheX** [comment]
有时候sft数据量有点大，希望支lazy-tokenize、streaming、packing、resume、workers之间不要有互斥，比如支持lazy-tokenize的同时也支持packing等等。

**Comment by howitry** [comment]
> > 希望可以增加对Omni模型talker部分的微调训练支持哈
> 
> 计划中 会在swift 4.1 支持

请问目前omini模型talker部分的audio codec是有开源实现吗？没有codec模型的话似乎没办法微调talker

**Comment by Jintao-Huang** [comment]
> 希望优化dataloader，sft 数据量不小，多节点训练时，目前发现:
> 
> 1. streaming  效率较低
>    2.使用前置 mapping dataset: interleave_prob调整后会有漫长的重新mapping过程

streaming模式确实效率不高，建议不开streaming。会使用dataloadershard

**Comment by Jintao-Huang** [comment]
> 有时候sft数据量有点大，希望支lazy-tokenize、streaming、packing、resume、workers之间不要有互斥，比如支持lazy-tokenize的同时也支持packing等等。

lazy_tokenize 和packing一起是没法支持的，因为不知道length就没法packing。不知道packing后的数据长度 就不知道每个epoch的train_iters

**Comment by Jintao-Huang** [comment]
streaming速度慢的问题，后面看看如何缓解

**Comment by wizyoung** [comment]
> streaming速度慢的问题，后面看看如何缓解

参考下 https://github.com/mosaicml/streaming?

**Comment by JoshonSmith** [comment]
> > > 希望可以增加对Omni模型talker部分的微调训练支持哈
> > 
> > 
> > 计划中 会在swift 4.1 支持
> 
> 请问目前omini模型talker部分的audio codec是有开源实现吗？没有codec模型的话似乎没办法微调talker

https://github.com/QwenLM/Qwen3-TTS 最新开源了，应该与qwen3-omni的Talker是相同的

**Comment by LLMOON** [comment]
希望swift-megatron后端也能接入mlflow进行模型版本管理

**Comment by Xhtshr** [comment]
希望支持最新的trl 0.27.0和vllm 0.14.0

**Comment by JoshonSmith** [comment]
support Qwen3-tts, Qwen3-asr finetune

**Comment by hcs129** [comment]
许愿：能否增加 megatron后端下的插件化功能， 除了reward之外，增加对 插件化 loss function的支持

**Comment by nicknign** [comment]
DeepSeek-V3.2 是否已支持？有计划支持的时间吗？

**Comment by LiXinYuECNU** [comment]
目前的swift在omni系列模型上的强化学习资料太少了
比如qwen3-omni的强化学习，很难跑起来，虽然有很多issue提出来了，但是根本不好解决
同样的swift rlhf的代码，把模型换成qwen2.5-omni就能正常跑，但3的moe就不行，建议还是统一下吧，Megatron的方式我试了，依然很难跑通qwen3-omni，可以对于这些模型训练的环境进行一个公开，这样方便复现和使用

**Comment by randydl** [comment]
新年好，请问现在可以使用main分支了吗，4.0版本的进度咋样啦，谢谢

**Comment by Jintao-Huang** [comment]
> 新年好，请问现在可以使用main分支了吗，4.0版本的进度咋样啦，谢谢

可以使用的，稳定的

**Comment by Jintao-Huang** [comment]
> DeepSeek-V3.2 是否已支持？有计划支持的时间吗？

明天会开始接入 deepseekv3.2/GLM5.0的模型

**Comment by Jintao-Huang** [comment]
swift4.0 scheduled for release: 2026.3.2

**Comment by SushantGautam** [comment]
Waiting for swift4.0 !!!

**Comment by shizhenjun** [comment]
支持一下多轮DPO吧，比如DMPO。

**Comment by JeremySun1224** [comment]
看到ms-swift将在4.1版本支持RAY，很期待。请问有大概日期么？

**Comment by Darrenzeng** [comment]
使用swift3.12的各个版本+dapo训练+zero3+80*80G显卡，都会出现OOM，训练模型为qwen3-32b。在swift3.12版本前没有出现过这种情况。

**Comment by hjh0119** [comment]
@Darrenzeng maybe fixed in https://github.com/modelscope/ms-swift/pull/8091

**Comment by wdndev** [comment]
megatron sft dpo等训练时，能否添加channel loss功能，channel loss功能挺实用的，但是在megatron训练时没有，希望能添加。

**Comment by PolarisHsu** [comment]
> swift4.0 scheduled for release: 2026.3.2

催更催更~

**Comment by Jintao-Huang** [comment]
ms-swift 4.0 has been released. Due to time constraints, the feature requests mentioned by users above will be gradually optimized in the next few versions. Thank you all.

release note: https://github.com/modelscope/ms-swift/releases/tag/v4.0.0

**Comment by KingArtherTT** [comment]
期望强化学习的 Rollout阶段，能支持 ray自主组装用于vLLM Server部署的机器。
比如：Megatron GRPO训练时，可以使用 ray 将4台8卡H20组装到一起，增加vLLM Server实例，从而显著降低 Rollout时的耗时，也能显著降低 GPU 气泡，尤其在较大模型的训练上。
（包括：基于当前权重进行采样、能透过 ray 来更新不同vLLM Server实例的权重等）

**Comment by Jintao-Huang** [comment]
> > DeepSeek-V3.2 是否已支持？有计划支持的时间吗？
> 
> 明天会开始接入 deepseekv3.2/GLM5.0的模型

deepseek v3.2: https://github.com/modelscope/ms-swift/pull/8226
GLM-5: https://github.com/modelscope/ms-swift/pull/8085

**Comment by gusario** [comment]
Would be cool to have working example with GRPO trained with usage of OpenEnv https://github.com/meta-pytorch/OpenEnv, TRL already has some examples here https://github.com/huggingface/trl/tree/main/examples/scripts/openenv

**Comment by HJXA** [comment]
请问4.0能否对packing后的数据进行缓存呢，我当前PT时，载入7亿样本数量后（tokenize过），往往需要packing几个小时

**Comment by li624120638** [comment]
4.0的mla在rope上的编码和3.0的出入很大，3.0的代码里面这么一段：
```
if multi_latent_attention:
        x1 = t[..., 0::2]
        x2 = t[..., 1::2]
        t = torch.cat((x1, x2), dim=-1)

# first part is cosine component
# second part is sine component, need to change signs with _rotate_half method
cos_ = (torch.cos(freqs) * mscale).to(t.dtype)
sin_ = (torch.sin(freqs) * mscale).to(t.dtype)

t = (t * cos_) + (_rotate_half(t, rotary_interleaved) * sin_)
return torch.cat((t, t_pass), dim=-1)
```
的`if multi_latent_attention`部分被删除了，取而代之的是在`_rotate_half`中进行interleave，但是在`t*cos`部分就缺少了对应的interleave，虽然在计算freqs也做了对应的interleave，但是感觉还是不太对劲

**Comment by Jintao-Huang** [comment]
> 4.0的mla在rope上的编码和3.0的出入很大，3.0的代码里面这么一段：
> 
> ```
> if multi_latent_attention:
>         x1 = t[..., 0::2]
>         x2 = t[..., 1::2]
>         t = torch.cat((x1, x2), dim=-1)
> 
> # first part is cosine component
> # second part is sine component, need to change signs with _rotate_half method
> cos_ = (torch.cos(freqs) * mscale).to(t.dtype)
> sin_ = (torch.sin(freqs) * mscale).to(t.dtype)
> 
> t = (t * cos_) + (_rotate_half(t, rotary_interleaved) * sin_)
> return torch.cat((t, t_pass), dim=-1)
> ```
> 
> 的`if multi_latent_attention`部分被删除了，取而代之的是在`_rotate_half`中进行interleave，但是在`t*cos`部分就缺少了对应的interleave，虽然在计算freqs也做了对应的interleave，但是感觉还是不太对劲

可以跑一下 https://github.com/modelscope/ms-swift/blob/b59a105ec77a974b1839175c9b416f6ddc7988f9/tests/megatron/test_align/test_llm.py#L97-L98

对齐是没问题的

参考这里：https://github.com/huggingface/transformers/blob/1f553bdc1703c78e272656ab8fb86d6494593e18/src/transformers/models/deepseek_v3/modeling_deepseek_v3.py#L444-L447

**Comment by Jintao-Huang** [comment]
主要是为了适配 deepseek-v3.2/glm-5做的修改。参考了transformers的rope实现。并做了精度对齐的测试。

**Comment by li624120638** [comment]
> 主要是为了适配 deepseek-v3.2/glm-5做的修改。参考了transformers的rope实现。并做了精度对齐的测试。

3.0中`multi_latent_attention=true`，`rotary_interleaved`默认为false，`t`在`t = (t * cos_) + (_rotate_half(t, rotary_interleaved) * sin_)`之前做了interlevae，在`_rotate_half`中没做，结果是：interleave的t与非interleave的cos，interleave的t与非interleave的sin进行计算
4.0中`multi_latent_attention=true`，`rotary_interleaved`默认为true，freq在初始时进行了interleave，导致cos和sin也interleave，但是取消了`multi_latent_attention=true`的interleave，结果是：非interleave的t与interleave的cos，interleave的t与interleave的sin进行计算。
即使强制在4.0中将`rotary_interleaved`设置为false，结果是：非interleave的t与非interleave的cos，非interleave的t与非interleave的sin计算，也和3.0版本对不上。不知道我这么理解对不对，实际代码跑下来，例如`q_pos_emb`在两个版本中的最大差值能达到9.多，
大佬说的精度对齐，是指和deepseek-v3.2/glm-5对齐，并不是和swfit3.0兼容吗？

**Comment by Jintao-Huang** [comment]
稍等 我重新核对一下。我看着也有点不对

精度对齐是 transformers forward与megatron forward输出的logits误差。

**Comment by Jintao-Huang** [comment]
精度对齐上确实是没问题的... 你要不单开一个issue吧
swift 4
<img width="888" height="50" alt="Image" src="https://github.com/user-attachments/assets/b56f271e-80f9-4623-b35b-29d27ed5fe66" />

swift 3
<img width="863" height="68" alt="Image" src="https://github.com/user-attachments/assets/25ab36e5-b585-4ce7-9df7-d99302188492" />

**Comment by Jintao-Huang** [comment]
> 稍等 我重新核对一下。我看着也有点不对
> 
> 精度对齐是 transformers forward与megatron forward输出的logits误差。

bugfix: https://github.com/modelscope/ms-swift/pull/8462

**Comment by Jintao-Huang** [comment]
> 请问4.0能否对packing后的数据进行缓存呢，我当前PT时，载入7亿样本数量后（tokenize过），往往需要packing几个小时

use cached_dataset https://github.com/modelscope/ms-swift/tree/main/examples/train/cached_dataset

and use packing_num_proc

**Comment by HJXA** [comment]
> > 请问4.0能否对packing后的数据进行缓存呢，我当前PT时，载入7亿样本数量后（tokenize过），往往需要packing几个小时
> 
> use cached_dataset https://github.com/modelscope/ms-swift/tree/main/examples/train/cached_dataset
> 
> and use packing_num_proc

谢谢回答，在这个问题中packing_num_proc是更关键的，其能大大提升速度

**Comment by Jintao-Huang** [comment]
ms-swift 4.2-4.4 roadmap: https://github.com/modelscope/ms-swift/issues/9028
